rules.go

```diff
@@ -0,0 +1,265 @@
+	Origin string
+	// a value of Referer header
+	Referer string
+	// Defines is request looks like XHLHttpRequest
+	IsXHR bool
+}
+
+const (
+	addressPart = iota
+	domainName
+	exactAddress
+	regexRule
+)
+
+type ruleAdBlock struct {
+	ruleText    string
+	regex       *regexp.Regexp
+	options     map[string]bool
+	isException bool
+	domains     map[string]bool
+	ruleType    int
+}
+
+func parseRule(ruleText string) (*ruleAdBlock, error) {
+	rule := &ruleAdBlock{
+		ruleText: strings.TrimSpace(ruleText),
+		domains:  map[string]bool{},
+		options:  map[string]bool{},
+	}
+
+	if rule.ruleText == "" {
+		return nil, ErrEmptyLine
+	}
+
+	if strings.HasPrefix(rule.ruleText, "!") || strings.HasPrefix(rule.ruleText, "[Adblock") {
+		return nil, ErrSkipComment
+	}
+
+	if strings.Contains(rule.ruleText, "##") || strings.Contains(rule.ruleText, "#@#") || strings.Contains(rule.ruleText, "#?#") {
+		return nil, ErrSkipHTML
+	}
+
+	rule.isException = strings.HasPrefix(rule.ruleText, "@@")
+	if rule.isException {
+		rule.ruleText = rule.ruleText[2:]
+	}
+
+	if strings.Contains(rule.ruleText, "$") {
+		parts := strings.SplitN(rule.ruleText, "$", 2)
+		rule.ruleText = parts[0]
+
+		for _, option := range strings.Split(parts[1], ",") {
+			optionNegative := !strings.HasPrefix(option, "~")
+			option = strings.TrimPrefix(option, "~")
+			_, supportedOption := supportedOptionsPat[option]
+
+			switch {
+			case strings.HasPrefix(option, "domain="):
+				for _, domain := range strings.Split(option[len("domain="):], "|") {
+					name := strings.TrimSpace(domain)
+					rule.domains[strings.TrimPrefix(name, "~")] = !strings.HasPrefix(name, "~")
+				}
+			case !supportedOption:
+				return nil, ErrUnsupportedRule
+			default:
+				rule.options[option] = optionNegative
+			}
+		}
+	}
+
+	rule.ruleType = addressPart
+	if strings.HasPrefix(rule.ruleText, "||") && strings.HasSuffix(rule.ruleText, "^") {
+		rule.ruleType = domainName
+	}
+
+	if strings.HasPrefix(rule.ruleText, "|") && strings.HasSuffix(rule.ruleText, "|") {
+		rule.ruleType = exactAddress
+	}
+
+	// The empty rule means the will block everything
+	// /{anything}/ mean regular expression. or define some other pattern to conflict to a path like /anything/
+	if rule.ruleText == "" || (strings.HasPrefix(rule.ruleText, "/") && strings.HasSuffix(rule.ruleText, "/")) {
+		rule.ruleType = regexRule
+	}
+
+	re, err := regexp.Compile(ruleToRegexp(rule))
+	if err != nil {
+		return nil, ErrCompilingRegex
+	}
+	rule.regex = re
+	return rule, nil
+}
+
+// RuleSet handle the structure to match whitelist and blacklist
+type RuleSet struct {
+	white *matcher
+	black *matcher
+}
+
+// Allow return of the current request is allowed to proceed or should be avoided
+func (ruleSet *RuleSet) Allow(req *Request) bool {
+	return ruleSet.white.Match(req) || !ruleSet.black.Match(req)
+}
+
+// NewRuleSetFromList creates a RuleSet from a list of rules
+func NewRuleSetFromList(rulesStr []string) (*RuleSet, error) {
```

Comment:
In real-world, we need a function, where we pass a list of parsed rules. Why so? We can do parsing in the application, manage errors (decide if we want to skip anything else, for example) there and pass ready rules to make a ruleset. Otherwise, usage would be really hard and clumsy. Let me give you an example:

```golang
func getAdblock(url string) (*adblockgoparser.RuleSet, error) {
    response, err := http.Get(url)
    if err != nil {
        if response != nil {
            _ = response.Body.Close()
        }
        return fmt.Errorf("cannot make a request for the adblock list: %w", err)
    }
    defer func() {
        io.Copy(ioutil.Discard, response.Body)
        response.Body.Close()
    }()

    scanner := bufio.NewScanner(response.Body)
    rules := []*adblockgoparser.Rule{}
    for scanner.Scan() {
        rule, err := adblockgoparser.Parse(scanner.Text())
        switch {
        case err == nil:
            rules = append(rules, rule)
        case errors.Is(adblockgoparser.ErrSkipComment), errors.Is(adblockgoparser.ErrUnsupportedRule), errors.Is(adblockgoparser.ErrEmptyLine):
        default:
            return fmt.Errorf("unknown error: %w", err)
        }
    }

    return rules, nil
}
```

Such code should be a part of the _application_, not this _library_. The application is responsible for how to react to errors. This function, for example, gives no possibility to skip comments.

That's for example why I'm asking to expose ParseRule function